### PR TITLE
Add optional offset to LogisticRegressionExpr

### PR DESCRIPTION
**Pull request by kianfar77**
Signed-off-by: kianfar77 <kiavash.kianfar@databricks.com>

## What changes are proposed in this pull request?

This PR adds an optional offset argument to the `logistic_regression_gwas` function. 

## How is this patch tested?
- [x] Unit tests

(Details)

### CODE REVIEW

File: core/src/main/scala/org/apache/spark/sql/catalyst/expressions/QuinaryExpression.scala [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+ */
+
+package org.apache.spark.sql.catalyst.expressions
+
+import org.apache.spark.sql.catalyst.InternalRow
+import org.apache.spark.sql.catalyst.expressions.codegen.Block._
+import org.apache.spark.sql.catalyst.expressions.codegen._
+
+/**
+ * An expression with four inputs + 5th optional input and one output.
```
Comment (henrydavidge) [inline core/src/main/scala/org/apache/spark/sql/catalyst/expressions/QuinaryExpression.scala:24]: nit: I think there's always one output. Might be confusing to specify one output when that's a property of all expressions.

File: core/src/main/scala/org/apache/spark/sql/catalyst/expressions/QuinaryExpression.scala [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+        $nullSafeEval""")
+    } else {
+      ev.copy(
+        code = code"""
+        ${firstGen.code}
+        ${secondGen.code}
+        ${thirdGen.code}
+        ${fourthGen.code}
+        ${fifthGen.map(_.code).getOrElse("")}
+        ${CodeGenerator.javaType(dataType)} ${ev.value} = ${CodeGenerator.defaultValue(dataType)};
```
Comment (henrydavidge) [inline core/src/main/scala/org/apache/spark/sql/catalyst/expressions/QuinaryExpression.scala:147]: We should have tests for these two lines

File: core/src/test/scala/io/projectglow/tertiary/LogisticRegressionSuite.scala [hunk 3]
```diff
@@ -463,4 +518,5 @@ class LogisticRegressionSuite extends GlowBaseTest {
       LogisticRegressionGwas.logitTests.get("FIRTH"))
     assert(LogisticRegressionGwas.logitTests.get("monkey").isEmpty)
   }
+
```
Comment (henrydavidge) [inline core/src/test/scala/io/projectglow/tertiary/LogisticRegressionSuite.scala:521]: nit: delete this

File: core/src/test/scala/io/projectglow/tertiary/LogisticRegressionSuite.scala [hunk 4]
```diff
@@ -89,93 +97,119 @@ class LogisticRegressionSuite extends GlowBaseTest {
...
+  private val admitStudentsWithZeroOffset: TestData = {
+    val zeroOffset = Array.fill[Double](50)(0)
+    admitStudentsWithOffset.copy(offsetOption = Some(zeroOffset))
   }
-
+  private val admitStudents: TestData = admitStudentsWithOffset.copy(offsetOption = None)
   private val admitStudentsStats =
     LogitTestResults(-0.611263, 0.54266503, Seq(2.901759e-01, 1.014851), 0.04693173)
 
+  // Data amd golden stats are from R's logistf package
```
Comment (henrydavidge) [inline core/src/test/scala/io/projectglow/tertiary/LogisticRegressionSuite.scala:181]: amd => and

File: core/src/main/scala/io/projectglow/functions.scala [hunk 5]
```diff
@@ -364,12 +364,18 @@ object functions {
...
    */
+  def logistic_regression_gwas(genotypes: Column, phenotypes: Column, covariates: Column, test: String, offset: Column): Column = withExpr {
+    new io.projectglow.sql.expressions.LogisticRegressionExpr(genotypes.expr, phenotypes.expr, covariates.expr, Literal(test), offset.expr)
+  }
+
   def logistic_regression_gwas(genotypes: Column, phenotypes: Column, covariates: Column, test: String): Column = withExpr {
     new io.projectglow.sql.expressions.LogisticRegressionExpr(genotypes.expr, phenotypes.expr, covariates.expr, Literal(test))
   }
 
+
```
Comment (henrydavidge) [inline core/src/main/scala/io/projectglow/functions.scala:378]: Where did this extra new line come from?

File: core/src/main/scala/io/projectglow/sql/expressions/FirthTest.scala [hunk 6]
```diff
@@ -85,8 +90,8 @@ object FirthTest extends LogitTest {
 
     while (!converged && !exploded && iter <= maxIter) {
       try {
-        args.mu := x(::, 0 until m0) * args.b
-        sigmoid.inPlace(args.mu)
+        val eta = offsetOption.fold(x(::, 0 until m0) * args.b)(_ + x(::, 0 until m0) * args.b)
```
Comment (henrydavidge) [inline core/src/main/scala/io/projectglow/sql/expressions/FirthTest.scala:93]: Prefer pattern matching or if else over fold as per https://github.com/databricks/scala-style-guide#monadic-chaining

File: functions.yml [hunk 7]
```diff
@@ -457,6 +462,9 @@ gwas_functions:
         - name: test
           doc: Which logistic regression test to use. Can be ``LRT`` or ``Firth``
           type: str
+        - name: offset
+          doc: An optional double array of offset values
```
Comment (henrydavidge) [inline functions.yml:466]: Can you briefly explain what the offset does?

**Review by henrydavidge — COMMENTED** [review]
Looks pretty good overall. Mostly minor comments.

File: core/src/main/scala/org/apache/spark/sql/catalyst/expressions/QuinaryExpression.scala (hunk 1 above)
Comment (kianfar77) [reply core/src/main/scala/org/apache/spark/sql/catalyst/expressions/QuinaryExpression.scala:24]: Just copied from OSS but I agree. Removed.

File: core/src/main/scala/org/apache/spark/sql/catalyst/expressions/QuinaryExpression.scala (hunk 2 above)
Comment (kianfar77) [reply core/src/main/scala/org/apache/spark/sql/catalyst/expressions/QuinaryExpression.scala:147]: Added QuinaryExpressionSuite.

File: core/src/test/scala/io/projectglow/tertiary/LogisticRegressionSuite.scala (hunk 3 above)
Comment (kianfar77) [reply core/src/test/scala/io/projectglow/tertiary/LogisticRegressionSuite.scala:521]: Done

File: core/src/test/scala/io/projectglow/tertiary/LogisticRegressionSuite.scala (hunk 4 above)
Comment (kianfar77) [reply core/src/test/scala/io/projectglow/tertiary/LogisticRegressionSuite.scala:181]: Fixed

File: core/src/main/scala/io/projectglow/sql/expressions/FirthTest.scala (hunk 6 above)
Comment (kianfar77) [reply core/src/main/scala/io/projectglow/sql/expressions/FirthTest.scala:93]: changed to pattern matching everywhere.

File: functions.yml (hunk 7 above)
Comment (kianfar77) [reply functions.yml:466]: Added

File: core/src/main/scala/io/projectglow/functions.scala (hunk 5 above)
Comment (kianfar77) [reply core/src/main/scala/io/projectglow/functions.scala]: It was an extra line we had in the template for the is_optional case. I removed it.

File: core/src/main/scala/io/projectglow/functions.scala [hunk 8]
```diff
@@ -166,7 +166,6 @@ object functions {
     new io.projectglow.sql.expressions.HardCalls(probabilities.expr, numAlts.expr, phased.expr)
   }
 
-
```
Comment (henrydavidge) [inline core/src/main/scala/io/projectglow/functions.scala:169]: Thanks for these changes

**Review by henrydavidge — APPROVED** [review]
Looks good to me, thanks!

**Pull request merged**